src/features/typeDefinitionProvider.ts

```diff
@@ -0,0 +1,98 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as serverUtils from '../omnisharp/utils';
+import { CancellationToken, TypeDefinitionProvider, Location, Position, TextDocument, Uri } from 'vscode';
+import { MetadataRequest, MetadataSource, GoToTypeDefinitionRequest, } from '../omnisharp/protocol';
+import { createRequest, toRange3, toVscodeLocation } from '../omnisharp/typeConversion';
+import AbstractSupport from './abstractProvider';
+import DefinitionMetadataOrSourceGeneratedDocumentProvider from './definitionMetadataDocumentProvider';
+import { OmniSharpServer } from '../omnisharp/server';
+import { LanguageMiddlewareFeature } from '../omnisharp/LanguageMiddlewareFeature';
+import SourceGeneratedDocumentProvider from './sourceGeneratedDocumentProvider';
+
+export default class CSharpTypeDefinitionProvider extends AbstractSupport implements TypeDefinitionProvider {
+    constructor(
+        server: OmniSharpServer,
+        private definitionMetadataDocumentProvider: DefinitionMetadataOrSourceGeneratedDocumentProvider,
+        private sourceGeneratedDocumentProvider: SourceGeneratedDocumentProvider,
+        languageMiddlewareFeature: LanguageMiddlewareFeature) {
+        super(server, languageMiddlewareFeature);
+    }
+
+    public async provideTypeDefinition(document: TextDocument, position: Position, token: CancellationToken): Promise<Location[]> {
+
+        let req = <GoToTypeDefinitionRequest>createRequest(document, position);
+        req.WantMetadata = true;
+
+        const locations: Location[] = [];
+        try {
+            const goToTypeDefinitionResponse = await serverUtils.goToTypeDefinition(this._server, req, token);
+            // the defintion is in source
+            if (goToTypeDefinitionResponse && goToTypeDefinitionResponse.Definitions) {
+
+                for (const definition of goToTypeDefinitionResponse.Definitions) {
```

Comment:
This logic should be sharable with https://github.com/OmniSharp/omnisharp-vscode/blob/090acf09c8/src/features/definitionProvider.ts#L36, pretty much in its entirety. Let's refactor these so they don't have to be maintained separately.

Reply:
I've updated to use the same `CSharpDefinitionProvider` class to implement both `DefinitionProvider` and `TypeDefinitionProvider`, and put as much as I could into a shared function.